dfastmi/batch/FileNameRetriever.py

```diff
@@ -27,121 +27,210 @@
+
+        Returns
+        -------
+        filenames : Dict[Any, Tuple[str,str]]
+            Dictionary of string tuples representing the D-Flow FM file names for
+            each reference/with measure pair.
+        """
+    
+    def _cfg_get(self, config: configparser.ConfigParser, chap: str, key: str) -> str:
+        """
+        Get a single entry from the analysis configuration structure.
+        Raise KeyError message when it fails.
+
+        Arguments
+        ---------
+        config : Optional[configparser.ConfigParser]
+            The variable containing the configuration (may be None for imode = 0).
+        chap : str
+            The name of the chapter in which to search for the key.
+        key : str
+            The name of the key for which to return the value.
+
+        Raises
+        ------
+        KeyError
+            If the key in the chapter doesn't exist.
+
+        Returns
+        -------
+        value : str
+            The value specified for the key in the chapter.
+        """
+        value = config.get(chap, key, fallback= None)
+        if value is None:
+            raise KeyError(f'Keyword "{key}" is not specified in group "{chap}" of analysis configuration file.')
+        return value
+
+class FileNameRetrieverFactory:
     """
-    Extract the list of 2N file names from the configuration.
-    This routine is valid for version 2 configuration files.
-
-    Arguments
-    ---------
-    needs_tide : bool
-        Specifies whether the tidal boundary is needed.
-
-    config : Optional[configparser.ConfigParser]
-        The variable containing the configuration (may be None for if 0).
-
-    Returns
-    -------
-    filenames : Dict[Any, Tuple[str,str]]
-        Dictionary of string tuples representing the D-Flow FM file names for
-        each reference/with measure pair. The keys of the dictionary vary. They
-        can be the discharge index, discharge value or a tuple of forcing
-        conditions, such as a Discharge and Tide forcing tuple.
+    Class is used to register and get creation of AFileNameRetriever Objects
     """
-    filenames: Dict[Any, Tuple[str,str]]
-    key: Union[Tuple[float, int], float]
-    filenames = {}
-    i = 0
-    while True:
-        i = i + 1
-        CSTR = "C{}".format(i)
-        if CSTR in config:
-            Q = float(cfg_get(config, CSTR, "Discharge"))
-            reference = cfg_get(config, CSTR, "Reference")
-            measure = cfg_get(config, CSTR, "WithMeasure")
-            if needs_tide:
-               T = cfg_get(config, CSTR, "TideBC")
-               key = (Q,T)
-            else:
-               key = Q
-            filenames[key] = (reference, measure)
+    _creators = {}
+    """Contains the AFileNameRetriever Objects creators to be used"""
+
+    def __init__(self):
+        self._creators = {}
+
+    def register_creator(self, version: str, creator: Callable[[bool], AFileNameRetriever]):
```

Comment:
May be use Version object instead of a str... 
str of "2.0" is not same as "2"... but i think version("2.0 ") == version("2")

Reply:
I tried to get it working before, but since the retrieved version from the config is a string it didnt work, will try it again by making them both version and see if I can get that to work.

Reply:
done